x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/io/stream/PlanNamedTypes.java

```diff
@@ -511,14 +514,45 @@ static void writeEsQueryExec(PlanStreamOutput out, EsQueryExec esQueryExec) thro
     }
 
     static EsSourceExec readEsSourceExec(PlanStreamInput in) throws IOException {
-        return new EsSourceExec(in.readSource(), readEsIndex(in), readAttributes(in), in.readOptionalNamedWriteable(QueryBuilder.class));
+        return new EsSourceExec(
+            in.readSource(),
+            readEsIndex(in),
+            readAttributes(in),
+            in.readOptionalNamedWriteable(QueryBuilder.class),
+            readIndexMode(in)
+        );
     }
 
     static void writeEsSourceExec(PlanStreamOutput out, EsSourceExec esSourceExec) throws IOException {
         out.writeNoSource();
         writeEsIndex(out, esSourceExec.index());
         writeAttributes(out, esSourceExec.output());
         out.writeOptionalNamedWriteable(esSourceExec.query());
+        writeIndexMode(out, esSourceExec.indexMode());
+    }
+
+    static IndexMode readIndexMode(StreamInput in) throws IOException {
+        if (in.getTransportVersion().onOrAfter(TransportVersions.ESQL_ADD_INDEX_MODE_TO_SOURCE)) {
+            int ord = in.readByte();
+            return switch (ord) {
+                case 0 -> IndexMode.STANDARD;
```

Comment:
Nit: make 0, 1 etc part of the enum and add checks to enforce they're distinct?

Reply:
I pushed https://github.com/elastic/elasticsearch/pull/108747/commits/b07cd89f31168ea9540ee1ee3f6ab84421b0b1f3 to serialize the name of the index mode instead.